### PR TITLE
 Add a new `Style/EvalWithLocation` cop

### PR DESCRIPTION
**Pull request by pocke**
`eval` can receive a file name and a line number to display actual backtrace.

For example:

```ruby
eval <<-RUBY
  raise rescue p $!.backtrace
RUBY
# => ["(eval):1:in `<main>'", "test.rb:1:in `eval'", "test.rb:1:in `<main>'"]

eval <<-RUBY, binding, __FILE__, __LINE__
  raise rescue p $!.backtrace
RUBY
# => ["test.rb:1:in `<main>'", "test.rb:1:in `eval'", "test.rb:1:in `<main>'"]
```

If we pass `__FILE__` and `__LINE__` to an `eval` method, an exception that occurs in the eval method has a backtrace with actual file name and line number. But sometimes we forget writing `__FILE__` and `__LINE__`.

This cop detects this problem. This cop adds an offense for an `eval` method call without a location.

Note: I think we can implement auto-correction for this cop in the next step.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests(`rake spec`) are passing.
* [x] The new code doesn't generate RuboCop offenses that are checked by `rake internal_investigation`.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/style/eval_with_location.rb [hunk 1]
```diff
@@ -0,0 +1,144 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # TODO
```
Comment (bbatsov) [inline lib/rubocop/cop/style/eval_with_location.rb]: I guess you forgot about this. :-)

File: lib/rubocop/cop/style/eval_with_location.rb [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      #   RUBY
+      #
+      #   # good
+      #   C.class_eval <<-RUBY, __FILE__, __LINE__ + 1
+      #     def do_something
+      #     end
+      #   RUBY
+      class EvalWithLocation < Cop
+        MSG = 'Pass `__FILE__` and `__LINE__` to `eval` method ' \
+              'to display actual backtrace.'.freeze
```
Comment (bbatsov) [inline lib/rubocop/cop/style/eval_with_location.rb]: I think this should be reworded to something mentioning "source location metadata (that's used by backtraces)."

File: lib/rubocop/cop/style/eval_with_location.rb [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        MSG_INCORRECT_LINE = 'Use `%<expected>s` instead of `%<actual>s` ' \
+                             'to display actual backtrace.'.freeze
```
Comment (bbatsov) [inline lib/rubocop/cop/style/eval_with_location.rb]: Same here.

File: lib/rubocop/cop/style/eval_with_location.rb (hunk 1 above)
Comment (pocke) [reply lib/rubocop/cop/style/eval_with_location.rb]: 🙀 I'll update it!

**Comment by pocke** [comment]
I added the documentation and updated the messages.

> I think this should be reworded to something mentioning "source location metadata (that's used by backtraces)."
 

Sorry, I'm not sure what you meant.



I updated the message to ``Pass `__FILE__` and `__LINE__` to `eval` method. That is used by backtraces.`` and ``Use `%<expected>s` instead of `%<actual>s`. That is used by backtraces.``
I think the messages should have `__FILE__` and `__LINE__`, because users can't understand how to fix this offense without these concrete metadata names. So I omit `source location metadata` from the messages.
What do you think?

**Comment by bbatsov** [comment]
I'd only replace the second sentence with `, as they are used by backtraces`. Generally we aim to have each message be a single sentence.

**Comment by bbatsov** [comment]
This has to be rebased, btw.

**Comment by pocke** [comment]
> I'd only replace the second sentence with , as they are used by backtraces. Generally we aim to have each message be a single sentence.


I understand. Thanks!

I'll update the messages, and rebase.

**Comment by bbatsov** [comment]
👍

**Comment by brandonweiss** [comment]
@pocke Why does the example in the PR pass `__LINE__` directly but the comment in the cop does `__LINE__ + 1`?

**Comment by pocke** [comment]
> @pocke Why does the example in the PR description pass __LINE__ directly but the comment in the cop does __LINE__ + 1?

It is my mistake, `__LINE__ + 1` is correct if you use a haredoc.

**Comment by brandonweiss** [comment]
🤔 I don't think I understand why the offset is necessary?

Perhaps the reasoning/instructions should be documented and/or included in part of the offense message? I imagine people will frequently look at the documentation and include the offset when they don't need to and/or not include the offset when they do need to.

**Comment by marcandre** [comment]
Wow, this is the first time that RuboCop actually signals a genuine error for me, thanks to this cop 🎉
Thanks @pocke 🙇‍♂️!

**Pull request merged**